src/nvim/fileio.c

```diff
@@ -5203,38 +5203,114 @@ static void vim_maketempdir(void)
   (void)umask(umask_save);
 }
 
+// Evaluate "expr" for readdir().
+static varnumber_T readdir_checkitem(typval_T *expr, const char *name)
+{
+  typval_T save_val;
+  typval_T rettv;
+  typval_T argv[2];
+  varnumber_T retval = 0;
+  bool error = false;
+
+  prepare_vimvar(VV_VAL, &save_val);
+  set_vim_var_string(VV_VAL, name, -1);
+  argv[0].v_type = VAR_STRING;
+  argv[0].vval.v_string = (char_u *)name;
+
+  if (eval_expr_typval(expr, argv, 1, &rettv) == FAIL) {
+    goto theend;
+  }
+
+  retval = tv_get_number_chk(&rettv, &error);
+  if (error) {
+    retval = -1;
+  }
+
+  tv_clear(&rettv);
+
+theend:
+  set_vim_var_string(VV_VAL, NULL, 0);
+  restore_vimvar(VV_VAL, &save_val);
+  return retval;
+}
+
+/// Core part of "readdir()" function.
+/// Retrieve the list of files/directories of "path" into "gap".
+void readdir_core(
+    garray_T *gap,
+    Directory *dirp,
+    const char *path,
+    typval_T *expr,
+    bool is_checkitem)
+{
+  ga_init(gap, (int)sizeof(char *), 20);
+
+  for (;;) {
+    bool ignore;
+
+    path = os_scandir_next(dirp);
```

Comment:
Can you use a local variable, `p`, instead of overriding `path` like in the vim patch?